### PR TITLE
feat(components): switch

### LINKED ISSUE #5417: Switch component
**Design:** [Figma](https://www.figma.com/file/Z6AGVUmKQzLNtDozFamW7f/s2-Mobile?node-id=40%3A1185)
Screenshot 2022-05-17 at 19 55 13

**Spec:** 
1. Recommended props
```tsx
type SwitchProps = {
	value: boolean,
	onChange: (value: boolean) => void
	isDisabled?: boolean,
}
```
2. Consider using https://reactnative.dev/docs/switch but you will probably need to create component from scratch because of limited styling
3. Don't forget to add animations

Maybe we can use Skia to create this component?

### LINKED ISSUE #9001: UI: Components - Switch
**Design:** [Figma](https://www.figma.com/file/Y2E8dcdNYrSMGIVRjYFXKp/%5Bs2.0%5D-Components?type=design&node-id=57%3A30408&mode=design&t=1Vnxs6y4zk5w2ndS-1)

**Spec:** [Notion](https://www.notion.so/satoshilabs/UI-Libraries-Visual-Reskin-a163f384aa6b41b49e76e3ebd6f3b637?pvs=4)

### PR DESCRIPTION
**Pull request by AdamSchinzel**
## Description

* Update the Switch component

## Related Issue

Resolve #9001

## Screenshots:

See Storybook

### CODE REVIEW

File: packages/components/src/components/form/Switch/Switch.tsx [hunk 1]
```diff
@@ -71,52 +70,80 @@ const CheckboxInput = styled.input`
...
-                type="checkbox"
-                role="switch"
-                checked={isChecked}
-                disabled={isDisabled}
-                onChange={handleChange}
-                aria-checked={isChecked}
-            />
-        </Container>
+        <Wrapper>
+            {labelPosition === 'before' && label && (
```
Comment (dahaca) [inline packages/components/src/components/form/Switch/Switch.tsx]: To not copy code and simplify margins in CSS, use `flex-direction` and `gap` :)

File: packages/components/src/components/form/Switch/Switch.tsx [hunk 2]
```diff
@@ -71,52 +70,80 @@ const CheckboxInput = styled.input`
...
+    text-align: left;
+`;
+
 export interface SwitchProps {
     id?: string;
     isChecked: boolean;
+    label?: string;
     onChange: (isChecked?: boolean) => void;
     isDisabled?: boolean;
-    isSmall?: boolean;
```
Comment (dahaca) [inline packages/components/src/components/form/Switch/Switch.tsx:78]: I believe this should be kept, for now, and marked as `TODO: legacy prop`. Will ask designers what to do with it.

File: packages/components/src/components/form/Switch/Switch.tsx [hunk 3]
```diff
@@ -71,52 +70,80 @@ const CheckboxInput = styled.input`
...
+    isAlert?: boolean;
     className?: string;
     dataTest?: string;
+    labelPosition?: 'before' | 'after';
```
Comment (dahaca) [inline packages/components/src/components/form/Switch/Switch.tsx]: nit: `left` and `right` would be clearer IMO, also used for `iconIlignment` in `Button`

File: packages/components/src/components/form/Switch/Switch.tsx [hunk 4]
```diff
@@ -71,52 +70,80 @@ const CheckboxInput = styled.input`
...
 `;
 
+const Label = styled.div<Pick<SwitchProps, 'labelPosition' | 'isDisabled' | 'isAlert'>>`
+    display: flex;
+    justify-content: center;
+    margin-left: ${({ labelPosition }) => (labelPosition === 'before' ? '0px' : '16px')};
+    margin-right: ${({ labelPosition }) => (labelPosition === 'after' ? '0px' : '16px')};
+    color: ${({ isAlert, isDisabled, theme }) =>
+        getLabelColor(theme, { alert: isAlert, disabled: isDisabled })};
+    font-size: ${FONT_SIZE.NORMAL};
```
Comment (dahaca) [inline packages/components/src/components/form/Switch/Switch.tsx]: use `typography` like [here](https://github.com/trezor/trezor-suite/blob/20b98a6f907855883c2ee811a28b9cb61ab9f997/packages/components/src/components/buttons/Button/Button.tsx#L60)

File: packages/components/src/config/colors.ts [hunk 5]
```diff
@@ -7,11 +7,12 @@ export const THEME = {
     light: {
         THEME: 'light',
 
-        BG_GREEN: '#00854D',
+        BG_GREEN: '#0F6148',
```
Comment (dahaca) [inline packages/components/src/config/colors.ts]: I wouldn't tough this file at all for now. Also, better start using colors from the `theme` package, the one used in Figma, check other updated components

File: packages/components/src/utils/utils.ts [hunk 6]
```diff
@@ -1,14 +1,25 @@
...
     if (!checked) {
         return theme.STROKE_GREY;
     }
-    return disabled ? darken(theme.DARKEN_20_PERCENT_FILTER, theme.STROKE_GREY) : theme.BG_GREEN;
+    return disabled ? theme.BG_GREEN_DISABLED : theme.BG_GREEN;
+};
+
+export const getLabelColor = (theme: DefaultTheme, { alert, disabled }: LabelColorOptions) => {
+    if (alert) {
+        return theme.STROKE_RED;
```
Comment (dahaca) [inline packages/components/src/utils/utils.ts]: use colors from the `theme` package 🙏

File: packages/components/src/components/form/Switch/Switch.tsx [hunk 7]
```diff
@@ -18,46 +24,39 @@ const Container = styled.div<Pick<SwitchProps, 'isChecked' | 'isDisabled' | 'isS
...
-
-            :active {
-                button {
-                    box-shadow: none;
-                }
+            :focus-within {
+                border: 1px solid ${theme.STROKE_BLUE};
+                background: ${isChecked ? theme.BG_GREEN : theme.STROKE_GREY};
+                box-shadow: 0px 0px 0px 3px ${({ theme }) => theme.BOX_SHADOW_SWITCH};
             }
```
Comment (dahaca) [inline packages/components/src/components/form/Switch/Switch.tsx]: I believe you can simply use [this](https://github.com/trezor/trezor-suite/blob/9cb5cc576820bd01ced4c5b0501605f3a529fe62/packages/components/src/components/buttons/Button/Button.tsx#L26). Feel free to move it to some other file as you see fit.

File: packages/components/src/components/form/Switch/Switch.tsx [hunk 8]
```diff
@@ -71,52 +70,80 @@ const CheckboxInput = styled.input`
...
-            isDisabled={isDisabled}
-            isSmall={isSmall}
-            onClick={e => {
-                e.preventDefault();
-                handleChange();
-            }}
-            className={className}
-            data-test={dataTest}
-        >
-            <Handle isSmall={isSmall} isChecked={isChecked} disabled={isDisabled} type="button" />
```
Comment (dahaca) [inline packages/components/src/components/form/Switch/Switch.tsx:109]: add `tabIndex={-1}` to the `Handle` so that only 1 element is focusable

File: packages/components/src/components/form/Switch/Switch.tsx [hunk 9]
```diff
@@ -18,46 +24,39 @@ const Container = styled.div<Pick<SwitchProps, 'isChecked' | 'isDisabled' | 'isS
...
-                }
+            :hover {
+                background: ${isChecked ? theme.BG_GREEN_HOVER : theme.STROKE_GREY_HOVER};
             }
-
-            :active {
-                button {
-                    box-shadow: none;
-                }
+            :focus-within {
```
Comment (dahaca) [inline packages/components/src/components/form/Switch/Switch.tsx]: put `:hover` after `:focus-within` so that it can be hovered even if focused.

Comment (dahaca) [inline packages/components/src/components/form/Switch/Switch.tsx]: `:focus-within:has(:focus-visible)` so that it doesn't gain the outline when the Handle is clicked

**Review by dahaca — CHANGES_REQUESTED** [review]

File: packages/components/src/components/form/Switch/Switch.tsx (hunk 1 above)
Comment (AdamSchinzel) [reply packages/components/src/components/form/Switch/Switch.tsx]: 755f9fbbae98c6194cb5d379ff90e4ea918ce5f3

File: packages/components/src/components/form/Switch/Switch.tsx (hunk 2 above)
Comment (AdamSchinzel) [reply packages/components/src/components/form/Switch/Switch.tsx:78]: 755f9fbbae98c6194cb5d379ff90e4ea918ce5f3

File: packages/components/src/components/form/Switch/Switch.tsx (hunk 3 above)
Comment (AdamSchinzel) [reply packages/components/src/components/form/Switch/Switch.tsx]: 755f9fbbae98c6194cb5d379ff90e4ea918ce5f3

File: packages/components/src/components/form/Switch/Switch.tsx (hunk 4 above)
Comment (AdamSchinzel) [reply packages/components/src/components/form/Switch/Switch.tsx]: 755f9fbbae98c6194cb5d379ff90e4ea918ce5f3

File: packages/components/src/config/colors.ts (hunk 5 above)
Comment (AdamSchinzel) [reply packages/components/src/config/colors.ts]: 755f9fbbae98c6194cb5d379ff90e4ea918ce5f3

File: packages/components/src/utils/utils.ts (hunk 6 above)
Comment (AdamSchinzel) [reply packages/components/src/utils/utils.ts]: 755f9fbbae98c6194cb5d379ff90e4ea918ce5f3

File: packages/components/src/components/form/Switch/Switch.tsx (hunk 8 above)
Comment (AdamSchinzel) [reply packages/components/src/components/form/Switch/Switch.tsx:109]: 755f9fbbae98c6194cb5d379ff90e4ea918ce5f3

File: packages/components/src/components/form/Switch/Switch.tsx (hunk 9 above)
Comment (AdamSchinzel) [reply packages/components/src/components/form/Switch/Switch.tsx]: 755f9fbbae98c6194cb5d379ff90e4ea918ce5f3

Comment (AdamSchinzel) [reply packages/components/src/components/form/Switch/Switch.tsx]: 755f9fbbae98c6194cb5d379ff90e4ea918ce5f3

File: packages/components/src/components/form/Switch/Switch.tsx (hunk 7 above)
Comment (AdamSchinzel) [reply packages/components/src/components/form/Switch/Switch.tsx]: After changes in the last commit I don't think if I can reuse that much there now

File: packages/components/src/components/form/Switch/Switch.tsx (hunk 1 above)
Comment (dahaca) [reply packages/components/src/components/form/Switch/Switch.tsx]: nit: use `spacingsPx` instead of pixel values when possible.

File: packages/components/src/components/form/Switch/Switch.tsx [hunk 10]
```diff
@@ -71,52 +80,79 @@ const CheckboxInput = styled.input`
     width: 1px;
 `;
 
+const Label = styled.div<Pick<SwitchProps, 'isDisabled' | 'isAlert' | 'isSmall'>>`
+    color: ${({ isAlert, isDisabled, theme }) =>
+        getLabelColor(theme, { alert: isAlert, disabled: isDisabled })};
+    text-align: left;
```
Comment (dahaca) [inline packages/components/src/components/form/Switch/Switch.tsx]: I believe `left` is the default value of `text-align`

File: packages/components/src/components/form/Switch/Switch.tsx [hunk 11]
```diff
@@ -71,52 +80,79 @@ const CheckboxInput = styled.input`
...
+    isAlert?: boolean;
+    isSmall?: boolean; // TODO: legacy prop
     className?: string;
     dataTest?: string;
+    labelPosition?: 'left' | 'right';
 }
 
 export const Switch = ({
     onChange,
     id,
```
Comment (dahaca) [inline packages/components/src/components/form/Switch/Switch.tsx]: i'd remove this prop and use `getWeakRandomId()` locally to get an ID and pass it to the label, what do you think? Then [this](https://github.com/trezor/trezor-suite/blob/00ce57306b73ada0c40c45d273fc77cfad5bae7b/packages/suite/src/views/wallet/send/components/Outputs/components/Amount/index.tsx#L230) would have to be refactored + you could try out the label prop ;)

File: packages/components/src/components/form/Switch/Switch.tsx [hunk 12]
```diff
@@ -71,52 +80,79 @@ const CheckboxInput = styled.input`
...
+    color: ${({ isAlert, isDisabled, theme }) =>
+        getLabelColor(theme, { alert: isAlert, disabled: isDisabled })};
+    text-align: left;
+    ${({ isSmall }) => (isSmall ? typography.label : typography.body)}
+`;
+
 export interface SwitchProps {
     id?: string;
     isChecked: boolean;
+    label?: string;
```
Comment (dahaca) [inline packages/components/src/components/form/Switch/Switch.tsx]: should be a `ReactNode`, woudl allow more flexibility + we are usually using the <Translation/> component

File: packages/components/src/components/form/Switch/Switch.tsx [hunk 13]
```diff
@@ -1,63 +1,72 @@
...
-            :focus-within {
-                button {
-                    box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 1);
-                }
+            :focus-within:has(:focus-visible) {
+                border: 1px solid ${theme.backgroundAlertBlueBold};
+                background: ${isChecked
+                    ? theme.backgroundPrimaryDefault
+                    : theme.backgroundNeutralDisabled};
+                box-shadow: ${MEDIA_QUERY.DARK_THEME
```
Comment (dahaca) [inline packages/components/src/components/form/Switch/Switch.tsx]:
```
  ${MEDIA_QUERY.DARK_THEME} {
                :focus-within:has(:focus-visible) {
                    box-shadow: ${boxShadows.focusedDark};
                }
            }

```

File: packages/components/src/components/form/Switch/Switch.tsx (hunk 7 above)
Comment (dahaca) [reply packages/components/src/components/form/Switch/Switch.tsx]: Try this, I think it's still nicer to have this unified:

```
export const getFocusShadowStyle = (selector = ':focus-visible') => css`
    border: 1px solid transparent;

    ${selector} {
        border-color: ${({ theme }) => theme.backgroundAlertBlueBold};
        box-shadow: ${boxShadows.focusedLight};
    }

    ${MEDIA_QUERY.DARK_THEME} {
        ${selector} {
            box-shadow: ${boxShadows.focusedDark};
        }
    }
`;
```

And then `getFocusShadowStyle(':focus-within:has(:focus-visible)')`, seems like it's working alright.

File: packages/components/src/components/form/Switch/Switch.stories.tsx [hunk 14]
```diff
@@ -1,26 +1,13 @@
...
-
-        return (
-            <SwitchComponent
-                onChange={handleIsChecked}
-                isChecked={isChecked}
-                isSmall={args.isSmall}
-                isDisabled={args.isDisabled}
-            />
-        );
+export const Switch: StoryObj = {
```
Comment (dahaca) [inline packages/components/src/components/form/Switch/Switch.stories.tsx]: Revert the change, please, it's not working on click like this :DD I also fell into that trap when reviewting yesterday.

And for nicer controls you can add:

```
argTypes: {
        labelPosition: {
            options: ['left', 'right'],
            control: {
                type: 'radio',
            },
        },
    },
```

**Review by dahaca — CHANGES_REQUESTED** [review]

File: packages/components/src/components/form/Switch/Switch.tsx [hunk 15]
```diff
@@ -1,63 +1,72 @@
...
+    width: ${({ isSmall }) => (isSmall ? '32px' : '44px')};
     flex-shrink: 0;
     margin: 0px;
     padding: 3px;
     position: relative;
     background: ${({ isChecked, isDisabled, theme }) =>
         getInputColor(theme, { checked: isChecked, disabled: isDisabled })};
-    border-radius: 12px;
+    border-radius: ${borders.radii.sm};
     transition: background 0.25s ease 0s;
```
Comment (dahaca) [inline packages/components/src/components/form/Switch/Switch.tsx]: also, please add ` border-color 0.1s ease-out, box-shadow 0.1s ease-out` :)

Comment (AdamSchinzel) [reply packages/components/src/components/form/Switch/Switch.tsx]: 9ada0fde38e5e788ad4b7eba036e50f07baef91f

File: packages/components/src/components/form/Switch/Switch.stories.tsx (hunk 14 above)
Comment (AdamSchinzel) [reply packages/components/src/components/form/Switch/Switch.stories.tsx]: 9ada0fde38e5e788ad4b7eba036e50f07baef91f

File: packages/components/src/components/form/Switch/Switch.tsx (hunk 13 above)
Comment (AdamSchinzel) [reply packages/components/src/components/form/Switch/Switch.tsx]: 9ada0fde38e5e788ad4b7eba036e50f07baef91f

File: packages/components/src/components/form/Switch/Switch.tsx (hunk 12 above)
Comment (AdamSchinzel) [reply packages/components/src/components/form/Switch/Switch.tsx]: 9ada0fde38e5e788ad4b7eba036e50f07baef91f

File: packages/components/src/components/form/Switch/Switch.tsx (hunk 11 above)
Comment (AdamSchinzel) [reply packages/components/src/components/form/Switch/Switch.tsx]: lmk if we understand right each other 9ada0fde38e5e788ad4b7eba036e50f07baef91f

File: packages/components/src/components/form/Switch/Switch.tsx (hunk 10 above)
Comment (AdamSchinzel) [reply packages/components/src/components/form/Switch/Switch.tsx]: 9ada0fde38e5e788ad4b7eba036e50f07baef91f

File: packages/components/src/components/form/Switch/Switch.tsx (hunk 7 above)
Comment (AdamSchinzel) [reply packages/components/src/components/form/Switch/Switch.tsx]: 9ada0fde38e5e788ad4b7eba036e50f07baef91f

File: packages/components/src/components/form/Switch/Switch.tsx (hunk 1 above)
Comment (AdamSchinzel) [reply packages/components/src/components/form/Switch/Switch.tsx]: 9ada0fde38e5e788ad4b7eba036e50f07baef91f

**Review by dahaca — APPROVED** [review]
